### PR TITLE
[SPMD] Change assertions to raising value errors in distributed.py

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55254 Clang-format distributed.py
* #55253 [SPMD] Restrict DDP communication hooks to SPSD mode
* #55212 [SPMD] Error out SPMD in C++ Reducer
* #54826 [SPMD] Remove test_grad_layout_1devicemodule_2replicaperprocess
* **#54825 [SPMD] Change assertions to raising value errors in distributed.py**
* #54454 [SPMD] Error out SPMD mode

These assertions are tested in test_c10d.py

Context: https://github.com/pytorch/pytorch/pull/54454#discussion_r602657818

Differential Revision: [D27381649](https://our.internmc.facebook.com/intern/diff/D27381649/)

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -395,18 +395,30 @@ def __init__(
...
+        if (
+            device_ids is None
+            or self.device_type == "cpu"
+            or self.is_multi_device_module
+        ):
+            if device_ids or output_device:
+                raise ValueError(
+                    "DistributedDataParallel device_ids and output_device arguments "
+                    "only work with single-device GPU modules, but got "
+                    "device_ids {}, output_device {}, and module parameters {}.".format(
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:416]: Unrelated to this PR, but should it say "...and multiple devices/CPU device for module parameters {p.device for p in module.parameters()}" ?

**Review by rohan-varma — APPROVED** [review]
Thanks! I think ValueError is better than assertions in a library.

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (SciPioneer) [reply torch/nn/parallel/distributed.py:416]: Yep, updated! Thanks!

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -396,10 +396,13 @@ def __init__(
 
         self.is_multi_device_module = len({p.device for p in module.parameters()}) > 1
         distinct_device_types = {p.device.type for p in module.parameters()}
-        assert len(distinct_device_types) == 1, (
-            "DistributedDataParallel's input module must be on "
-            "the same type of devices, but input module parameters locate in {}."
-        ).format(distinct_device_types)
+        if len(distinct_device_types) != 1:
```
Comment (pritamdamania87) [inline torch/nn/parallel/distributed.py:399]: There is an assert on line 389 as well and another one in `_passing_sync_batchnorm_handle`.

**Pull request closed without merging**